File: Blind 75/15 - Greedy/55-jump-game.py
'''
You are given an integer array nums. You are initially positioned at the array's first index, and each element in the array represents your maximum jump length at that position.

Return true if you can reach the last index, or false otherwise.
'''
from time import time
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        target = len(nums) - 1
        # Start at second last element
        for i in range(len(nums) - 2, -1, -1):
            if i + nums[i] >= target:
                target = i

        return target == 0

    def reference(self, nums: List[int]) -> bool:
        goal = len(nums) - 1

        for i in range(len(nums) - 2, -1, -1):
            if i + nums[i] >= goal:
                goal = i
        return goal == 0

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.canJump(case))
                else:
                    self.canJump(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [2, 3, 1, 1, 4],
        [3, 2, 1, 0, 4],
        # Additional
        [0, 2, 3],
    ]
    test.quantify(test_cases)

File: Blind 75/15 - Greedy/JavaScript/55-jump-game.js
const { performance } = require("perf_hooks");

class Solution {
  method(nums) {
    let target = nums.length - 1;
    for (let i = nums.length - 2; i >= 0; i--) {
      if (i + nums[i] >= target) target = i;
    }

    return target === 0;
  }

  reference(nums, right = nums.length - 1) {
    for (let i = right; 0 <= i; i--) {
      const isJumpable = right <= i + nums[i];
      if (isJumpable) right = i;
    }

    return right === 0;
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [2, 3, 1, 1, 4],
  [3, 2, 1, 0, 4],
  // Additional
  [0, 2, 3],
];
test.quantify(testCases);
